File: plugins/custom_native_video_player_ios/android/src/main/kotlin/me/albemala/native_video_player/platform_interface/NativeVideoPlayerApi.kt
package me.albemala.native_video_player.platform_interface

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

interface NativeVideoPlayerApiDelegate {
    fun loadVideoSource(videoSource: VideoSource)
    fun getVideoInfo(): VideoInfo
    fun getPlaybackPosition(): Int
    fun play()
    fun pause()
    fun stop()
    fun isPlaying(): Boolean
    fun seekTo(position: Int)
    fun setPlaybackSpeed(speed: Double)
    fun setVolume(volume: Double)
}

const val invalidArgumentsErrorCode = "invalid_argument"
const val invalidArgumentsErrorMessage = "Invalid arguments"

class NativeVideoPlayerApi(
    messenger: BinaryMessenger,
    viewId: Int
) : MethodChannel.MethodCallHandler {

    var delegate: NativeVideoPlayerApiDelegate? = null

    private val channel: MethodChannel

    init {
        val name = "me.albemala.native_video_player.api.${viewId}"
        channel = MethodChannel(messenger, name)
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
    }

    fun onPlaybackReady() {
        channel.invokeMethod("onPlaybackReady", null)
    }

    fun onPlaybackEnded() {
        channel.invokeMethod("onPlaybackEnded", null)
    }

    fun onError(error: Error) {
        channel.invokeMethod("onError", error.message)
    }

    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when (methodCall.method) {
            "loadVideoSource" -> {
                val videoSourceAsMap = methodCall.arguments as? Map<*, *>
                    ?: return result.error(invalidArgumentsErrorCode, invalidArgumentsErrorMessage, null)
                val videoSource = VideoSource.from(videoSourceAsMap)
                    ?: return result.error(invalidArgumentsErrorCode, invalidArgumentsErrorMessage, null)
                delegate?.loadVideoSource(videoSource)
                result.success(null)
            }
            "getVideoInfo" -> {
                result.success(delegate?.getVideoInfo()?.toMap())
            }
            "getPlaybackPosition" -> {
                result.success(delegate?.getPlaybackPosition())
            }
            "play" -> {
                delegate?.play()
                result.success(null)
            }
            "pause" -> {
                delegate?.pause()
                result.success(null)
            }
            "stop" -> {
                delegate?.stop()
                result.success(null)
            }
            "isPlaying" -> {
                val playing = delegate?.isPlaying()
                result.success(playing)
            }
            "seekTo" -> {
                val position = methodCall.arguments as? Int
                    ?: return result.error(invalidArgumentsErrorCode, invalidArgumentsErrorMessage, null)
                delegate?.seekTo(position)
                result.success(null)
            }
            "setPlaybackSpeed" -> {
                val speed = methodCall.arguments as? Double
                    ?: return result.error(invalidArgumentsErrorCode, invalidArgumentsErrorMessage, null)
                delegate?.setPlaybackSpeed(speed)
                result.success(null)
            }
            "setVolume" -> {
                val volume = methodCall.arguments as? Double
                    ?: return result.error(invalidArgumentsErrorCode, invalidArgumentsErrorMessage, null)
                delegate?.setVolume(volume)
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}

File: plugins/custom_native_video_player_ios/ios/Classes/PlatformInterface/NativeVideoPlayerApi.swift
protocol NativeVideoPlayerApiDelegate: AnyObject {
    func loadVideoSource(videoSource: VideoSource)
    func getVideoInfo() -> VideoInfo
    func getPlaybackPosition() -> Int
    func play()
    func pause()
    func stop(completion: @escaping () -> Void)
    func isPlaying() -> Bool
    func seekTo(position: Int, completion: @escaping () -> Void)
    func setPlaybackSpeed(speed: Double)
    func setVolume(volume: Double)
}

let invalidArgumentsFlutterError = FlutterError(code: "invalid_argument", message: "Invalid arguments", details: nil)

class NativeVideoPlayerApi {
    weak var delegate: NativeVideoPlayerApiDelegate?

    let channel: FlutterMethodChannel

    init(
        messenger: FlutterBinaryMessenger,
        viewId: Int64
    ) {
        let name = "me.albemala.native_video_player.api.\(viewId)"
        channel = FlutterMethodChannel(
            name: name,
            binaryMessenger: messenger
        )
        channel.setMethodCallHandler(handleMethodCall)
    }

    deinit {
        channel.setMethodCallHandler(nil)
    }

    func onPlaybackReady() {
        channel.invokeMethod("onPlaybackReady", arguments: nil)
    }

    func onPlaybackEnded() {
        channel.invokeMethod("onPlaybackEnded", arguments: nil)
    }

    func onError(_ error: Error) {
        channel.invokeMethod("onError", arguments: error.localizedDescription)
    }

    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadVideoSource":
            guard let videoSourceAsMap = call.arguments as? [String: Any] else {
                result(invalidArgumentsFlutterError)
                return
            }
            guard let videoSource = VideoSource(from: videoSourceAsMap) else {
                result(invalidArgumentsFlutterError)
                return
            }
            delegate?.loadVideoSource(videoSource: videoSource)
            result(nil)
        case "getVideoInfo":
            let videoInfo = delegate?.getVideoInfo().toMap()
            result(videoInfo)
        case "getPlaybackPosition":
            let playbackPosition = delegate?.getPlaybackPosition()
            result(playbackPosition)
        case "play":
            delegate?.play()
            result(nil)
        case "pause":
            delegate?.pause()
            result(nil)
        case "stop":
            delegate?.stop {
                result(nil)
            }
        case "isPlaying":
            let playing = delegate?.isPlaying()
            result(playing)
        case "seekTo":
            guard let position = call.arguments as? Int else {
                result(invalidArgumentsFlutterError)
                return
            }
            delegate?.seekTo(position: position) {
                result(nil)
            }
        case "setPlaybackSpeed":
            guard let playbackSpeed = call.arguments as? Double else {
                result(invalidArgumentsFlutterError)
                return
            }
            delegate?.setPlaybackSpeed(speed: playbackSpeed)
            result(nil)
        case "setVolume":
            guard let volume = call.arguments as? Double else {
                result(invalidArgumentsFlutterError)
                return
            }
            delegate?.setVolume(volume: volume)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
